### DISCUSSION TITLE
Inmensa reducción del tiempo de carga de una página

### DISCUSSION THREAD

**Discussion by RodrigoTomeES** [elpatronaco/pccomponentes-buy-bot#15] [Ideas]
Buenas,

He estado evaluando que parte del código es la que más tiempo de CPU consume y por lo que he visto el crear la página y cargarla es la parte que mayor carga supone para la CPU. Pero no creo que se pueda mejorar el código que crea una nueva página en Chrome, pero si se puede mejorar y muchisimo el tiempo que tarda en realizar las peticiones.

Se puede lograr una mejora de casi aproximadamente 6 veces mejor bloqueando todo contenido innecesario al cargar la página de esta manera:

```javascript
await page.setRequestInterception(true)

page.on('request', (request) => {
    if (request.resourceType() === 'document') {
        request.continue()
    } else {
        request.abort()
    }
})
```

Pero no se si al bloquear los JS hace que no se pueda llevar a cabo la compra, por eso si alguno sabe como filtart los JS (para quitar todos los de analitica) o desactivar la intercepción justo antes de la compra.

Alguna idea @elpatronaco ?

**Comment by elpatronaco** [comment]
Has hecho pruebas ya con el snippet este que me pasas comprando en cualquier tienda ? Entiendo por donde quieres ir, pero filtrar solo las peticiones de html y dejar fuera todo lo demás rompería el funcionamiento de la página yo creo. Dime si te funciona y lo miro

**Reply by RodrigoTomeES** [reply]
> 
> 
> Has hecho pruebas ya con el snippet este que me pasas comprando en cualquier tienda ? Entiendo por donde quieres ir, pero filtrar solo las peticiones de html y dejar fuera todo lo demás rompería el funcionamiento de la página yo creo. Dime si te funciona y lo miro

He hecho pruebas y pasan cosas muy raras, si lo pones en debug y carga la interfaz es capaz de comprar por ejemplo en PCcomponentes pero si lo pones en modo headless da un null al buscar el botón del metodo de pago transacción.

Por esa razón, creo que habría que buscar un modo de desactivarlo justo al ir a pagar o en su defecto cargar una nueva página para comprar sin esta intercepción.

**Comment by elpatronaco** [comment]
Creo que lo que más recursos consume es el hecho de tener que cargar páginas enteras para cada gráfica, si estás en un ordena normalillo y pones pocas gráficas no tendrás problema, pero si pones 10 o más se complica. Creo q buscar stock se tendría que hacer con un scraper sencillo, no por navegador, sabes? Voy a probar con pccomponentes por ejemplo, a ver que tal saldría

**Reply by RodrigoTomeES** [reply]
> 
> 
> Creo que lo que más recursos consume es el hecho de tener que cargar páginas enteras para cada gráfica, si estás en un ordena normalillo y pones pocas gráficas no tendrás problema, pero si pones 10 o más se complica. Creo q buscar stock se tendría que hacer con un scraper sencillo, no por navegador, sabes? Voy a probar con pccomponentes por ejemplo, a ver que tal saldría

Sí, así es como funcionan la mayoría de bots que no compran hacen un wget de la página y buscan en el descargado el stock y si eso avisan pero claro pierdes la funcionalidad de comprar.

Creo que interceptando el request se puede obtener una gran mejora y mantener la opción de compra.

PD: tal y como está ahora el bot me consume más de 9GB de RAM (eso cuando mire ayer a la noche hoy estaría consumiendo más) y ha llegado a colpasar mi PC dejandolo sin señal de vídeo 😆 y no tengo un pc malillo es un ryzen 3600 + 32GB de RAM 😅

**Reply by elpatronaco** [reply]
Al diablo JAJAJA 32gb de ram y peta, es peor de lo q parecía entonces, nada nada dejémonos de interceptar cosas. Estoy trasteando con una librería que se llama scrape-it, que tiene una sintaxis muy muy sencilla y funciona por peticiones https, no tiene una página abierta para cada gráfica. Tarda algo mas de 1 segundo, imagino q es el tiempo que tarda en recuperar el html
![image](https://user-images.githubusercontent.com/38383754/111868366-d440d980-8979-11eb-8d43-bab409ba4b0d.png)

**Reply by RodrigoTomeES** [reply]
Tiene muy buena pinta la verdad, si se hace asi el scraping será mucho más eficiente. Si a eso le sumamos la precarga del navegador y la apertura de una página antes de iniciar el scraping seguramente la compra sea muy rápida porque solo tiene que redirigir la página a la url donde hay stock y comprar. Es una mejora brutal la verdad.

Si necesitas ayuda dime!, muchas gracias por el curro, a ver si consigo pillar una 3070 por menos de 700 😄

**Comment by elpatronaco** [comment]
Mírate la nueva rama :)

**Comment by elpatronaco** [comment]
Si pudieras acabar las otras tiendas me harías un favor la vd, me iba a comer ahora. Faltaría el scraping de coolmod y aussar y el script de aussar creo que está roto, pero bueno te lo digo por si te apetecía hacerlo, si no da igual

**Reply by RodrigoTomeES** [reply]
Me miro la rama a ver, que como fuera y me voy a ir en nada, a ver si me da tiempo a hacer algo

**Reply by RodrigoTomeES** [reply]
No he podido hacer mucho que me tengo que ir ya pero te lo he dejado en un pull request

**Reply by RodrigoTomeES** [reply]
@elpatronaco increible trabajo va volado en mi pi4 (solo un 30% de carga, antes llegaba al 100% y no iba), gracias crack 🥳

**Reply by elpatronaco** [reply]
@RodrigoTomeES de nada crack, gracias a ti!
